src/water_cmd.cpp

```diff
@@ -452,10 +454,15 @@ CommandCost CmdBuildCanal(TileIndex tile, DoCommandFlag flags, uint32 p1, uint32
 	TileArea ta(tile, p1);
 
 	/* Outside the editor you can only drag canals, and not areas */
-	if (_game_mode != GM_EDITOR && ta.w != 1 && ta.h != 1) return CMD_ERROR;
+	if (_game_mode != GM_EDITOR) {
+		TileArea ta(tile, p1);
+		if (ta.w != 1 && ta.h != 1) return CMD_ERROR;
+	}
 
 	CommandCost cost(EXPENSES_CONSTRUCTION);
-	TILE_AREA_LOOP(tile, ta) {
+	TileIterator* iter = HasBit(p2, 2) ? (TileIterator*)new DiagonalTileIterator(tile, p1) : new OrthogonalTileIterator(tile, p1);
```

Comment:
This feels... weird. What's that cast doing there? Also, codestyle - `TileIterator *iter`

Reply:
Codestyle fixed.

I copied this code directly from JGR's patchpack commit, but the same cast is done in [landscape.cpp](https://github.com/OpenTTD/OpenTTD/blob/master/src/landscape.cpp#L754), rail_cmd, and terraform_cmd, which are all the diagonal tile selectors I can think of (not counting terraforming, which is vertices).

Reply:
The cast should definitely not be necessary.